clients/src/main/java/org/apache/kafka/clients/consumer/CooperativeStickyAssignor.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.clients.consumer;
+
+import static org.apache.kafka.clients.consumer.CooperativeStickyAssignor.CooperativeUserData.decode;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import org.apache.kafka.common.protocol.types.Field;
+import org.apache.kafka.common.protocol.types.Schema;
+import org.apache.kafka.common.protocol.types.SchemaException;
+import org.apache.kafka.common.protocol.types.Struct;
+import org.apache.kafka.common.protocol.types.Type;
+
+/**
+ * A cooperative version of the {@link StickyAssignor StickyAssignor}. This follows the same (sticky)
+ * assignment logic as {@link StickyAssignor StickyAssignor} but allows for cooperative rebalancing. To turn on
+ * cooperative incremental rebalancing you must set all your consumers to use this {@code PartitionAssignor},
+ * or implement a custom one that returns {@code RebalanceProtocol.COOPERATIVE} in
+ * {@link CooperativeStickyAssignor#supportedProtocols supportedProtocols()}.
+ * <p>
+ * IMPORTANT: if upgrading from 2.3 or earlier, you must follow a specific upgrade path in order to safely turn on
+ * cooperative rebalancing. See the upgrade-guide for details
+ */
+public class CooperativeStickyAssignor extends StickyAssignor {
+
+    @Override
+    public List<RebalanceProtocol> supportedProtocols() {
+        return Arrays.asList(RebalanceProtocol.EAGER, RebalanceProtocol.COOPERATIVE);
+    }
+
+    @Override
+    public String name() {
+        return "cooperative-sticky";
+    }
+
+    @Override
+    Subscription buildSubscription(Set<String> topics) {
+        return new Subscription(
+            new ArrayList<>(topics), new CooperativeUserData(generation()).encode(), memberAssignment());
+    }
+
+    @Override
+    StickyUserData subscriptionToStickyUserData(Subscription subscription) {
+        CooperativeUserData userData = decode(subscription.userData());
+        return new StickyUserData(subscription.ownedPartitions(), Optional.of(userData.generation));
+    }
+
+    static final class CooperativeUserData {
```

Comment:
Meta: since we are likely only going to evolve this class in the future, I'd suggest we refactor the code a bit in the following way:

1) `StickyUserData` moved into this class, which contains version / generation / partitions (and future fields will be added here as well). Also the encode / decode function as-is.

2) Let `StickyAssignor extends CooperativeStickAssignor` and not vice versa, also use `StickyUserData` but define it's own schema, and overriding the encode / decode to skip the version information since it does not have it.

Then in the future, `StickAssignor`  class would not need to be updated and all enhancements goes to this class?

Reply:
I see your point about pushing the logic into `CooperativeStickyAssignor` since that's what we'll likely develop from here, but after playing around with this today I'm not sure if this is quite right. We want to evolve the functionality and/or userdata format of `CooperativeStickyAssignor` independently so it doesn't make sense to make `StickyAssignor` directly extend it. Most likely we'd have to force in some awkward case handling until we eventually had two pretty different classes. 
Given that, I think the best approach is extracting this out into an `AbstractStickyAssignor` -- WDYT?
Looking forward it might even be possible to further abstract the "balance vs stickiness" logic such that Streams could leverage it somehow -- but that's out of the current scope

Reply:
Playing it around a bit as well and now I think that makes sense (assuming the `StickyUserData` would be moved into this abstract class as well)?